File: core/src/main/java/com/toocol/ssh/core/mosh/core/network/Connection.java
package com.toocol.ssh.core.mosh.core.network;

import com.toocol.ssh.core.mosh.core.crypto.Crypto;
import com.toocol.ssh.utilities.utils.Timestamp;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.datagram.DatagramSocket;

import static com.toocol.ssh.core.mosh.core.network.NetworkConstants.*;

/**
 * Equivalent to network.h/network.cc/
 *
 * @author ：JoeZane (joezane.cn@gmail.com)
 * @date: 2022/5/11 22:54
 * @version: 0.0.1
 */
public final class Connection {

    final Transport.Addr addr;
    final DatagramSocket socket;
    final Crypto.Session encryptSession;
    final Crypto.Session decryptSession;
    final MoshPacket sendPacket;
    final Crypto.Message sendMessage;
    final MoshPacket receivePacket;
    final Crypto.Message receiveMessage;
    private short savedTimestamp;
    private long savedTimestampReceivedAt;
    private long expectedReceiverSeq;
    private long lastHeard;
    private long lastPortChoice;
    private long lastRoundtripSuccess;

    public Connection(Transport.Addr addr, DatagramSocket socket) {
        Crypto.Base64Key key = new Crypto.Base64Key(addr.key());
        this.addr = addr;
        this.socket = socket;
        this.encryptSession = new Crypto.Session(key);
        this.decryptSession = new Crypto.Session(key);
        this.sendPacket = new MoshPacket();
        this.sendMessage = new Crypto.Message();
        this.receivePacket = new MoshPacket();
        this.receiveMessage = new Crypto.Message();
    }

    public void send(byte[] msg) {
        MoshPacket packet = newPacket(msg);
        socket.send(Buffer.buffer(encryptSession.encrypt(packet.fillMessage(sendMessage))), addr.port(), addr.serverHost());
    }

    public byte[] recvOne(byte[] recv) {
        Crypto.Message decryptMessage = decryptSession.decrypt(recv, recv.length, receiveMessage);
        MoshPacket packet = receivePacket.resetData(decryptMessage);
        expectedReceiverSeq = packet.getSeq() + 1;
        if (packet.getTimestamp() != -1) {
            savedTimestamp = packet.getTimestamp();
            savedTimestampReceivedAt = Timestamp.timestamp();
        }
        return packet.getPayload();
    }

    public long timeout() {
        long rto = (long) Math.ceil(SRIT + 4 * RTTVAR);
        if (rto < MIN_RTO) {
            rto = MIN_RTO;
        } else if (rto > MAX_RTO) {
            rto = MAX_RTO;
        }
        return rto;
    }

    private MoshPacket newPacket(byte[] msg) {
        short outgoingTimestampReply = -1;

        long now = Timestamp.timestamp();

        if (now - savedTimestampReceivedAt < 1000) {
            outgoingTimestampReply = (short) (savedTimestamp + (short) (now - savedTimestampReceivedAt));
            savedTimestamp = -1;
            savedTimestampReceivedAt = -1;
        }

        return sendPacket.resetData(
                msg,
                MoshPacket.Direction.TO_SERVER,
                Timestamp.timestamp16(),
                outgoingTimestampReply
        );
    }

}

File: mosh/src/network/connection.rs
#![allow(dead_code)]
use std::{net::UdpSocket, time::Duration};
use utilities::TimeStamp;
use crate::crypto::{Base64Key, Session};
use super::{Direction, MoshPacket, MAX_RTO, MIN_RTO, RTTVAR, SRIT};
use log::info;

pub struct Connection {
    socket: UdpSocket,
    local_addr: String,
    remote_addr: String,

    session: Session,

    saved_timestamp: i64,
    saved_timestamp_receive_at: i64,
    expected_receiver_seq: u64,
}

impl Connection {
    const DIRECTION: Direction = Direction::ToServer;

    pub fn new(ip: &str, port: &str, key: &str) -> Self {
        let local_ip = local_ipaddress::get().expect("Get local ip address failed.");
        println!("local ip: {}, port: {}, key: {}", local_ip, port, key);
        let local_addr = format!("{}:{}", local_ip, port);
        let remote_addr = format!("{}:{}", ip, port);

        let socket = UdpSocket::bind(local_addr.as_str())
            .expect(format!("Bind to local udp socket failed, {}", local_addr.as_str()).as_str());
        socket.connect(remote_addr.as_str()).expect(
            format!(
                "Connect to remote udp socket failed, {}",
                remote_addr.as_str()
            )
            .as_str(),
        );
        socket
            .set_read_timeout(Some(Duration::from_millis(5)))
            .expect("Socket set read timeout failed.");

        Connection {
            socket,
            local_addr,
            remote_addr,
            session: Session::new(Base64Key::new(key.to_string())),
            saved_timestamp: -1,
            saved_timestamp_receive_at: 0,
            expected_receiver_seq: 0,
        }
    }

    pub fn send(&self, bytes: Vec<u8>) {
        self.socket.send(&bytes).expect(
            format!(
                "Udp socket send data failed, local_addr = {}, remote_addr = {}",
                self.local_addr, self.remote_addr
            )
            .as_str(),
        );
    }

    pub fn recv(&self) -> Option<Vec<u8>> {
        let mut recvd = vec![];
        if let Ok(len) = self.socket.recv(&mut recvd) {
            info!("Udp socket data received, len = {}", len);
            Some(recvd)
        } else {
            None
        }
    }

    pub fn recv_one(&mut self, bytes: Vec<u8>) -> Vec<u8> {
        let decrypt_message = self.session.decrypt(&bytes, bytes.len());
        let packet = MoshPacket::from_message(decrypt_message);
        self.expected_receiver_seq = packet.seq() + 1;
        self.saved_timestamp = packet.timestamp() as i64;
        self.saved_timestamp_receive_at = TimeStamp::timestamp() as i64;
        packet.payload()
    }

    pub fn timeout(&self) -> u64 {
        let mut rto = (SRIT + 4. * RTTVAR).ceil() as u64;
        if rto < MIN_RTO {
            rto = MIN_RTO
        } else if rto > MAX_RTO {
            rto = MAX_RTO
        }
        rto
    }

    fn new_packet(&mut self, msg: Vec<u8>) -> MoshPacket {
        let mut outgoing_timestamp_reply = -1i16;

        let now = TimeStamp::timestamp() as i64;

        if now - self.saved_timestamp_receive_at < 1000 {
            outgoing_timestamp_reply =
                self.saved_timestamp as i16 + (now as i16 - self.saved_timestamp_receive_at as i16);
            self.saved_timestamp = -1;
            self.saved_timestamp_receive_at = -1;
        }

        MoshPacket::from_payload(
            msg,
            Direction::ToServer,
            TimeStamp::timestamp_16(),
            outgoing_timestamp_reply as u16,
        )
    }
}
